src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,193 @@
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                //Check for a misplaced '\n'
+                if (line.Length <= 1 || line[line.Length - 2] != ByteCR)
+                {
+                    throw new FormatException("There was an issue with the frame format");
+                }
+
+                //Strip the \r\n from the span
+                line = line.Slice(0, line.Length - 2);
+
+                switch (_internalParserState)
+                {
+                    case InternalParseState.ReadMessageType:
+                        messageType = GetMessageType(line);
+                        start = lineEnd;
+                        _internalParserState = InternalParseState.ReadMessagePayload;
+                        consumed = lineEnd;
+
+                        break;
+                    case InternalParseState.ReadMessagePayload:
+                        if (!StartsWithDataPrefix(line))
+                        {
+                            throw new FormatException("Expected the message prefix 'data: '");
+                        }
+
+                        //Slice away the 'data: '
+                        var newData = line.Slice(6).ToArray();
+
+                        start = lineEnd;
+                        _data.Add(newData);
+                        consumed = lineEnd;
+                        break;
+                    case InternalParseState.ReadEndOfMessage:
+                        if (ReadCursorOperations.Seek(start, end, out lineEnd, ByteLF) == -1)
+                        {
+                            // The message has ended with \r\n\r
+                            return ParseResult.Incomplete;
+                        }
+
+                        if (_data.Count > 0)
+                        {
+                            //Find the final size of the payload
+                            var payloadSize = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                payloadSize += dataLine.Length;
+                            }
+                            var payload = new byte[payloadSize];
+
+                            //Copy the contents of the data array to a single buffer
+                            var offset = 0;
+                            foreach (var dataLine in _data)
+                            {
+                                dataLine.CopyTo(payload, offset);
+                                offset += dataLine.Length;
+                            }
+
+                            message = new Message(payload, messageType);
+                        }
+                        else
+                        {
+                            //Empty message
+                            message = new Message(Array.Empty<byte>(), messageType);
+                        }
+
+                        consumed = buffer.End;
```

Comment:
`consumed = lineEnd` right? You may not have reached the end of the entire buffer.

A good test for this would be to take a buffer with something like `data: T\r\ndata: Foo\r\n\r\ndata: T`, call Parse once and make sure that you haven't consumed the start of the second frame. I think with this code, you will.